codec/src/main/java/io/netty/handler/codec/compression/CompressionHandler.java

```diff
@@ -0,0 +1,115 @@
+ */
+public class CompressionHandler implements ChannelHandler {
+
+    private final Supplier<? extends Compressor> compressorSupplier;
+    private Compressor compressor;
+
+    /**
+     * Creates a new instance.
+     *
+     * @param compressorSupplier  the {@link Supplier} that is used to create the {@link Compressor}.
+     */
+    public CompressionHandler(Supplier<? extends Compressor> compressorSupplier) {
+        this.compressorSupplier = Objects.requireNonNull(compressorSupplier, "compressorSupplier");
+    }
+
+    @Override
+    public void handlerAdded(ChannelHandlerContext ctx) throws Exception {
+        compressor = compressorSupplier.get();
+    }
+
+    @Override
+    public void handlerRemoved(ChannelHandlerContext ctx) throws Exception {
+        if (compressor != null) {
+            finish(ctx, false);
+            compressor.close();
+            compressor = null;
+        }
+    }
+
+    @Override
+    public void channelInactive(ChannelHandlerContext ctx) throws Exception {
+        if (compressor != null) {
+            compressor.close();
+            compressor = null;
+        }
+        ctx.fireChannelInactive();
+    }
+
+    @Override
+    public Future<Void> write(ChannelHandlerContext ctx, Object msg) {
+        if (compressor == null || compressor.isFinished() || !(msg instanceof ByteBuf)) {
+            return ctx.write(msg);
+        }
+        ByteBuf input = (ByteBuf) msg;
+        try {
+            ByteBuf buffer = compressor.compress(input, ctx.alloc());
+            return ctx.write(buffer);
+        } finally {
+            input.release();
+        }
+    }
+
+    @Override
+    public Future<Void> close(ChannelHandlerContext ctx) {
+        return finish(ctx, true);
+    }
+
+    private Future<Void> finish(ChannelHandlerContext ctx, boolean closeCtx) {
+        if (compressor == null || compressor.isFinished()) {
+            if (closeCtx) {
+                return ctx.close();
+            }
+            return ctx.newSucceededFuture();
+        }
+        ByteBuf buffer = compressor.finish(ctx.alloc());
+        if (!buffer.isReadable()) {
+            buffer.release();
+            if (closeCtx) {
+                return ctx.close();
+            }
+            return ctx.newSucceededFuture();
+        }
+        if (closeCtx) {
+            Promise<Void> promise = ctx.newPromise();
+            Future<Void> f = ctx.writeAndFlush(buffer).addListener(ctx, (c, ignore) -> c.close().cascadeTo(promise));
+            if (!f.isDone()) {
+                // Ensure the channel is closed even if the write operation completes in time.
+                Future<?> sF =  ctx.executor().schedule(() -> ctx.close().cascadeTo(promise),
+                        10, TimeUnit.SECONDS); // FIXME: Magic number
+                f.addListener(sF, (scheduledFuture, ignore) -> scheduledFuture.cancel());
+            }
+            return promise.asFuture();
+        }
+        return ctx.write(buffer);
```

Comment:
Should this also be flushing like we are doing in the above `if` block?

Reply:
I don't think so as the flush is only needed to ensure the close is actually happen in a timeline manner (due the timeout handling).